src/network/ClientConnectionManager.ts

```diff
@@ -0,0 +1,804 @@
+            } else {
+                error = new IOError('No connection found to cluster since the client is starting.');
+            }
+        } else if (this.reconnectMode === ReconnectMode.ASYNC) {
+            error = new ClientOfflineError();
+        } else {
+            error = new IOError('No connection found to cluster.');
+        }
+        return error;
+    }
+
+    public getActiveConnections(): ClientConnection[] {
+        return Array.from(this.activeConnections.values());
+    }
+
+    public isAlive(): boolean {
+        return this.alive;
+    }
+
+    public getClientUuid(): UUID {
+        return this.clientUuid;
+    }
+
+    public getOrConnect(address: Address): Promise<ClientConnection> {
+        if (!this.client.getLifecycleService().isRunning()) {
+            return Promise.reject(new ClientNotActiveError('Client is not active.'));
+        }
+
+        const connection = this.getConnectionFromAddress(address);
+        if (connection) {
+            return Promise.resolve(connection);
+        }
+
+        const addressKey = address.toString();
+        const pendingConnection = this.pendingConnections.get(addressKey);
+        if (pendingConnection) {
+            return pendingConnection.promise;
+        }
+
+        const connectionResolver: Promise.Resolver<ClientConnection> = DeferredPromise<ClientConnection>();
+        this.pendingConnections.set(addressKey, connectionResolver);
+
+        const processResponseCallback = (msg: ClientMessage) => {
+            this.client.getInvocationService().processResponse(msg);
+        };
+
+        this.translate(address)
+            .then((translatedAddress) => {
+                if (translatedAddress == null) {
+                    connectionResolver.reject(new RangeError(`Address Translator could not translate address ${address}`));
+                    return;
+                }
+
+                this.triggerConnect(translatedAddress)
+                    .then((socket) => {
+                        const clientConnection = new ClientConnection(this.client, translatedAddress, socket,
+                            this.connectionIdCounter++);
+                        return this.initiateCommunication(socket)
+                            .then(() => {
+                                return clientConnection.registerResponseCallback(processResponseCallback);
+                            })
+                            .then(() => {
+                                return this.authenticateOnCluster(clientConnection);
+                            }).then(() => {
+                                connectionResolver.resolve(clientConnection);
+                            });
+                    })
+                    .catch((error) => {
+                        connectionResolver.reject(error);
+                    })
+                    .finally(() => {
+                        this.pendingConnections.delete(addressKey);
```

Comment:
nit: this `finally()` callback seems to duplicate the one on line 281. Is it on purpose?

Reply:
Yes. If this promise does not resolve in the specified timeout, we should delete the connection from the pending map. That's why I added the duplicate delete statement.

Reply:
In this method `connectionResolver` deferred promise depends on the promise chain started in `translate()`. If the latter promise chain gets resolved/rejected, `connectionResolver`'s `finally` callback should be triggered. So, that's why it seems to me that we can get rid of the first `finally` call and leave only the `connectionResolver`'s one. Am I missing something here?

Reply:
Oh, you are right. I removed the duplicate code.